cmd/integtest/main.go

```diff
@@ -0,0 +1,197 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+	"syscall"
+	"time"
+)
+
+const (
+	pgport = "12345"
+)
+
+// This command assumes it has free rein over $HOME/integration.
+var (
+	home     = os.Getenv("HOME")
+	dir      = home + "/integration"
+	lockFile = dir + "/lock"
+	wkdir    = dir + "/work"
+	gobin    = dir + "/bin"
+	pgdir    = dir + "/pg"
+	pgrun    = dir + "/pgrun" // for socket file
+)
+
+var (
+	flagT = flag.Duration("t", 15*time.Minute, "abort the test after the given duration")
+)
+
+func main() {
+	ctx := context.Background()
+	log.SetPrefix("integration: ")
+	log.SetFlags(log.Lshortfile | log.LstdFlags)
+	flag.Usage = usage
+	flag.Parse()
+	lock() // ensure only one at a time
+
+	ctx, cancel := context.WithTimeout(ctx, *flagT)
+	defer cancel()
+
+	if s := pgbin(); s != "" {
+		must(os.Setenv("PATH", os.Getenv("PATH")+":"+s))
+	}
+
+	must(os.RemoveAll(wkdir))
+	must(os.RemoveAll(gobin))
+	must(os.RemoveAll(pgdir))
+
+	if flag.NArg() < 1 {
+		usage()
+	}
+	pkg := flag.Arg(0)
+	args := flag.Args()[1:]
+
+	// accumulate environment for the test process
+	var env []string
+
+	if os.Getenv("CHAIN") == "" {
+		env = append(env, "CHAIN="+home+"/go/src/chain")
+	}
+
+	setupDB(ctx)
+	pgURL := "postgresql:///postgres?host=" + pgrun + "&port=" + pgport
+	env = append(env, "DB_URL_TEST="+pgURL) // for chain/database/pg/pgtest
+
+	buildTest(ctx, pkg)
+
+	must(os.MkdirAll(wkdir, 0700))
+
+	_, base := path.Split(pkg)
+	fmt.Println(base, strings.Join(args, " "))
+	cmd := command(ctx, gobin+"/"+base, args...)
+	cmd.Dir = wkdir
+	cmd.Env = mergeEnvLists(env, os.Environ())
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	err := cmd.Start()
+	if err != nil {
+		log.Printf("%s: %v", base, err)
+		panic("cmd failed")
+	}
+	defer syscall.Kill(-cmd.Process.Pid, syscall.SIGKILL)
+	err = cmd.Wait()
+	if err != nil {
+		log.Printf("%s: %v", base, err)
+		os.Exit(1)
+	}
+}
+
+func setupDB(ctx context.Context) {
+	fmt.Println("initdb", "-D", pgdir)
+	err := command(ctx, "initdb", "-D", pgdir).Run()
+	if err, ok := err.(*exec.ExitError); ok {
+		os.Stderr.Write(err.Stderr)
+	}
+	if err != nil {
+		log.Printf("initdb: %v", err)
+		panic("cmd failed")
+	}
+
+	var buf bytes.Buffer
+	must(configTemplate.Execute(&buf, map[string]string{
+		"port":    pgport,
+		"sockdir": pgrun,
+	}))
+	must(ioutil.WriteFile(pgdir+"/postgresql.conf", buf.Bytes(), 0600))
+
+	buf.Reset()
+	must(hbaTemplate.Execute(&buf, nil))
+	must(ioutil.WriteFile(pgdir+"/pg_hba.conf", buf.Bytes(), 0600))
+
+	fmt.Println("postgres", "-D", pgdir)
+	cmd := command(ctx, "postgres", "-D", pgdir)
```

Comment:
Do we always print the command before executing it? maybe `command` should print it to avoid accidental drift between the printlns and the executed commands?